201-front-door-api-management/azuredeploy.json

```diff
@@ -0,0 +1,1048 @@
+          {
+            "name": "subnets",
+            "count": "[length(parameters('virtualNetwork').subnets)]",
+            "input": {
+              "name": "[parameters('virtualNetwork').subnets[copyIndex('subnets')].name]",
+              "properties": {
+                "addressPrefix": "[parameters('virtualNetwork').subnets[copyIndex('subnets')].addressPrefix]",
+                "networkSecurityGroup": {
+                  "id": "[resourceId('Microsoft.Network/networkSecurityGroups', parameters('virtualNetwork').subnets[copyIndex('subnets')].nsg)]"
+                }
+              }
+            }
+          }
+        ]
+      }
+    },
+    {
+      "name": "[parameters('appInsightsName')]",
+      "type": "Microsoft.Insights/components",
+      "apiVersion": "2015-05-01",
+      "location": "[parameters('location')]",
+      "kind": "web",
+      "tags": {
+        "name": "[parameters('appInsightsName')]",
+        "resource": "[parameters('apimServiceName')]",
+        "service": "Application Insights"
+      },
+      "properties": {
+        "Application_Type": "web",
+        "Request_Source": "rest"
+      }
+    },
+    {
+      "type": "Microsoft.ApiManagement/service/loggers",
+      "name": "[concat(parameters('apimServiceName'), '/', parameters('appInsightsName'))]",
+      "apiVersion": "2019-01-01",
+      "dependsOn": [
+        "[variables('apimId')]"
+      ],
+      "properties": {
+        "loggerType": "applicationInsights",
+        "description": "Logger resources to APIM",
+        "credentials": {
+          "instrumentationKey": "[reference(variables('appInsightsId'), '2015-05-01').InstrumentationKey]"
+        }
+      }
+    },
+    {
+      "name": "[parameters('apimServiceName')]",
+      "type": "Microsoft.ApiManagement/service",
+      "apiVersion": "2019-01-01",
+      "location": "[parameters('location')]",
+      "tags": {
+        "name": "[parameters('apimServiceName')]",
+        "service": "APIM"
+      },
+      "sku": {
+        "name": "[parameters('apimSku')]",
+        "capacity": "[parameters('apimCapacity')]"
+      },
+      "identity": {
+        "type": "SystemAssigned"
+      },
+      "dependsOn": [
+        "[variables('vnetId')]"
+      ],
+      "properties": {
+        "publisherEmail": "[parameters('apimPublisherEmail')]",
+        "publisherName": "[parameters('apimPublisherName')]",
+        "virtualNetworkType": "External",
+        "hostnameConfigurations": [
+        ],
+        "additionalLocations": [
+        ],
+        "virtualNetworkConfiguration": {
+          "subnetResourceId": "[variables('apimSubnetId')]"
+        },
+        "certificates": [
+        ],
+        "customProperties": {
+        }
+      },
+      "resources": [
+        {
+          "apiVersion": "2017-03-01",
+          "type": "products",
+          "name": "custom",
+          "dependsOn": [
+            "[variables('apimId')]",
+            "[concat(variables('apimId'), '/apis/mockApi')]",
```

Comment:
remove concat and use resourceId

Reply:
Solved. I replaced these resource ids with the following variables:

"mockApiId": "[resourceId('Microsoft.ApiManagement/service/apis', parameters('apimServiceName'), 'mockApi')]",
        "postmanEchoApiId": "[resourceId('Microsoft.ApiManagement/service/apis', parameters('apimServiceName'), 'postmanEchoApi')]",
        "customProductId": "[resourceId('Microsoft.ApiManagement/service/products', parameters('apimServiceName'), 'custom')]",
        "testMethodId": "[resourceId('Microsoft.ApiManagement/service/apis/operations', parameters('apimServiceName'), 'mockApi', 'test')]"